libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/utils/RoutesUpdateReasonHelper.kt

```diff
@@ -0,0 +1,117 @@
+     * (or **steps' names**, see [isSameRoute]) are the same of primary routes
+     * of [legacyRoutes] and [newRoutes] (except **origin** coordinate) -> reason is [RoutesExtra.ROUTES_UPDATE_REASON_NEW_ROUTES];
+     * - if [RouteOptions.coordinatesList] are same for primary routes of [legacyRoutes] and
+     * [newRoutes] (except **origin** coordinate) -> reason is [RoutesExtra.ROUTES_UPDATE_REASON_NEW_ROUTES_ALTERNATIVE];
+     * - otherwise -> reason is [RoutesExtra.ROUTES_UPDATE_REASON_NEW_ROUTES].
+     *
+     * Note: function must be invoked before set a new route to NN, because it relies on
+     * off-route events.
+     */
+    @RoutesExtra.RoutesUpdateReason
+    internal fun getReason(
+        legacyRoutes: List<DirectionsRoute>,
+        newRoutes: List<DirectionsRoute>
+    ): String {
+        // most frequent cases
+        when {
+            newRoutes.isEmpty() ->
+                return RoutesExtra.ROUTES_UPDATE_REASON_CLEAN_UP
+            legacyRoutes.isEmpty() && newRoutes.isNotEmpty() ->
+                return RoutesExtra.ROUTES_UPDATE_REASON_NEW_ROUTES
+        }
+
+        val isSameRoutes = sameRoutes(legacyRoutes, newRoutes)
+        val isSameUuid = sameUuid(legacyRoutes, newRoutes)
+        val isSameRouteOptionsCoordinates = sameRouteOptionsCoordinates(legacyRoutes, newRoutes)
+        return when {
+            offRoute && isSameRouteOptionsCoordinates ->
+                RoutesExtra.ROUTES_UPDATE_REASON_NEW_ROUTES_REROUTE
+            isSameUuid && isSameRoutes ->
+                RoutesExtra.REASON_UPDATE_ROUTES_REFRESH
+            isSameRouteOptionsCoordinates ->
+                RoutesExtra.ROUTES_UPDATE_REASON_NEW_ROUTES_ALTERNATIVE
+            else -> RoutesExtra.ROUTES_UPDATE_REASON_NEW_ROUTES
+        }
+    }
+
+    private fun sameUuid(
+        legacyRoutes: List<DirectionsRoute>,
+        newRoutes: List<DirectionsRoute>
+    ): Boolean =
+        newRoutes.firstOrNull()?.isSameUuid(legacyRoutes.firstOrNull()) ?: false
+
+    private fun sameRoutes(
+        legacyRoutes: List<DirectionsRoute>,
+        newRoutes: List<DirectionsRoute>
+    ): Boolean =
+        newRoutes.firstOrNull()?.isSameRoute(legacyRoutes.firstOrNull()) ?: false
+
+    /**
+     * If a primary route has the same route options' coordinates - than **reason** should be
+     * re-route or alternative route.
+     */
+    private fun sameRouteOptionsCoordinates(
+        legacyRoutes: List<DirectionsRoute>,
+        newRoutes: List<DirectionsRoute>
+    ): Boolean =
+        ifNonNull(
+            legacyRoutes.firstOrNull()?.routeOptions()?.coordinatesList(),
+            newRoutes.firstOrNull()?.routeOptions()?.coordinatesList()
+        ) { legacyCoordinates, newCoordinates ->
+            val reversedLegacyCoordinates = legacyCoordinates.reversed()
+            val reversedNewCoordinates = newCoordinates.reversed()
+
+            for (idx in 0 until reversedNewCoordinates.lastIndex) { // until lastIndex-> skip Origin
```

Comment:
```suggestion
            for (idx in 0 until reversedNewCoordinates.lastIndex) {
```

Reply:
I would keep that note, it's not obvious why skipped the latest item